cloud/pkg/admissioncontroller/controller/admission.go

```diff
@@ -0,0 +1,273 @@
+
+// AdmissionController implements the admission webhook for validation of configuration.
+type AdmissionController struct {
+	Client *kubernetes.Clientset
+}
+
+func strPtr(s string) *string { return &s }
+
+// Start starts the webhook service
+func (ac *AdmissionController) Start(context *utils.CertContext) {
+	//TODO: read somewhere to get what's kind of webhook is enabled, register those webhook only.
+	err := ac.registerWebhookForDeviceModel(constants.WebhookNameDeviceModel, context)
+	if err != nil {
+		log.LOGGER.Fatalf("Failed to register the webhook with error: %v", err)
+	}
+	ac.deployService()
+	tlsConfig := utils.ConfigTLS(context)
+	http.HandleFunc("/devicemodels", serveDeviceModel)
+	server := &http.Server{
+		Addr:      fmt.Sprintf(":%v", constants.Port),
+		TLSConfig: tlsConfig,
+	}
+
+	server.ListenAndServeTLS("", "")
+}
+
+// Register registers the admission webhook.
+func (ac *AdmissionController) registerWebhookForDeviceModel(WebhookName string, context *utils.CertContext) error {
+	webhook := &admissionregistrationv1beta1.ValidatingWebhookConfiguration{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: WebhookName,
+		},
+		Webhooks: []admissionregistrationv1beta1.ValidatingWebhook{
+			{
+				Name: WebhookName,
+				Rules: []admissionregistrationv1beta1.RuleWithOperations{{
+					Operations: []admissionregistrationv1beta1.OperationType{
+						admissionregistrationv1beta1.Create,
+						admissionregistrationv1beta1.Update,
+						admissionregistrationv1beta1.Delete,
+						admissionregistrationv1beta1.Connect,
+					},
+					Rule: admissionregistrationv1beta1.Rule{
+						APIGroups:   []string{"devices.kubeedge.io"},
+						APIVersions: []string{"v1alpha1"},
+						Resources:   []string{"devicemodels"},
+					},
+				}},
+				ClientConfig: admissionregistrationv1beta1.WebhookClientConfig{
+					Service: &admissionregistrationv1beta1.ServiceReference{
+						Namespace: constants.NamespaceName,
+						Name:      constants.ServiceName,
+						Path:      strPtr("/devicemodels"),
+						Port:      pointer.Int32Ptr(constants.Port),
+					},
+					CABundle: context.SigningCert,
+				},
+			},
+		},
+	}
+
+	if err := ac.Client.AdmissionregistrationV1beta1().ValidatingWebhookConfigurations().Delete(webhook.Name, nil); err != nil {
+		serr, ok := err.(*errors.StatusError)
+		if !ok || serr.ErrStatus.Code != http.StatusNotFound {
+			log.LOGGER.Warnf("Could not delete existing webhook configuration: %v", err)
+		}
+	}
+
+	_, err := ac.Client.AdmissionregistrationV1beta1().ValidatingWebhookConfigurations().Create(webhook)
+	return err
+}
+
+func (ac *AdmissionController) deployService() {
+	service, err := ac.Client.CoreV1().Services(constants.NamespaceName).Get(constants.ServiceName, metav1.GetOptions{})
+	if err == nil && service != nil {
+		log.LOGGER.Infof("Service %s is already been created.", constants.ServiceName)
+		return
+	}
+	localIP := utils.GetIPv4Addr()
+	if localIP == "" {
+		log.LOGGER.Error("Cannot get one local valid IP")
```

Comment:
I'm not sure about this, but I would prefer to stop the creation of the service when this go wrong. What are you thinking?

Reply:
if it's not a valid IP, I think the service will fail to be created itself, this need to verify.

Reply:
Here is the error if it is not valid IP, 
```
Failed to create endpoints with error: Endpoints "webhook-service" is invalid: [subsets[0].addresses[0].ip: Invalid value: "": must be a valid IP address, (e.g. 10.9.8.7), subsets[0].addresses[0].ip: Invalid value: "": must be a valid IP address]
```